### ISSUE TITLE
Clustered vertx blocking vertx.eventBus().send

### ISSUE THREAD

**Issue by gkozyryatskyy** [eclipse/vert.x#1987]
**vertx-core** version 3.4.1
**vertx-hazelcast** version 3.4.1

Here is a little example
```
import io.vertx.core.AbstractVerticle;
import io.vertx.core.DeploymentOptions;
import io.vertx.core.Vertx;
import io.vertx.core.VertxOptions;

public class MainTest {

    public static void main(String[] args) throws Exception {
        VertxOptions options = new VertxOptions();
        deploy(Vertx.vertx(options));
        // Vertx.clusteredVertx(options, res -> {
        // deploy(res.result());
        // });
    }

    public static void deploy(Vertx vertx) {
        vertx.deployVerticle(new ReceiveVerticle(), new DeploymentOptions().setInstances(1).setWorker(true), e -> {
            // when ReceiveVerticle is deployed, deploy SendVerticle
            vertx.deployVerticle(new SendVerticle(), new DeploymentOptions().setInstances(1).setWorker(true));
        });
    }

    public static class SendVerticle extends AbstractVerticle {
        @Override
        public void start() throws Exception {
            // start executing some job
            for (int i = 0; i < 5; i++) {
                // send some partial result asynchronously to ReceiveVerticle
                String message = String.valueOf(i);
                System.out.println("Sending message " + message);
                vertx.eventBus().send("receive_address", message);
                Thread.sleep(1000l);
            }
        }
    }

    public static class ReceiveVerticle extends AbstractVerticle {
        @Override
        public void start() throws Exception {
            vertx.eventBus().consumer("receive_address", e -> {
                System.out.println("Receive message " + e.body());
            });
        }
    }

}
```
Here is and input of this example:
```
Sending message 0
Receive message 0
Sending message 1
Receive message 1
Sending message 2
Receive message 2
Sending message 3
Receive message 3
Sending message 4
Receive message 4
```
Everything is totally fine in non clustered vertx instance.
But when I do the same with clustered vertx
```
        // deploy(Vertx.vertx(options));
        Vertx.clusteredVertx(options, res -> {
            deploy(res.result());
        });
```
I have a totally different result
```
Sending message 0
Sending message 1
Sending message 2
Sending message 3
Sending message 4
Receive message 0
Receive message 1
Receive message 2
Receive message 3
Receive message 4
```
Messages have been sent just after SendVerticle execution finish. Seems like clustered vertx blocking `eventBus().send` from sending messages.

So the **question** is: How can I achieve in clustered vertx exactly the same result as in non clustered? I need this messages to be send asynchronously.

P.S. Also the interesting thing is that: If I use `.setMultiThreaded(true)` to the `SendVerticle.class` everything is fine... But it seems like a hack and may be there is a better way to achieve what I need because it is not working exactly in clustered vertx instance. And also my `SendVerticle` is really not multithreaded.

**Comment by gkozyryatskyy** [comment]
Ok. When I was looking for answers I found https://github.com/vert-x3/issues/issues/75 and https://github.com/vert-x3/vertx-hazelcast/pull/54. It seems like it is exactly the same problem and it was already fixed/merged. But it is still not working =(

**Comment by vietj** [comment]
we decided to not make this change after all and accept this is a restriction of sending message from execute blocking.

**Comment by gkozyryatskyy** [comment]
@vietj You mean if I will send message from `vertx.executeBlocking` with clustered vertx it will work exactly like with non clustered?

I check it and it is still not working.

Here is an example: 
```
public static void main(String[] args) throws Exception {
        VertxOptions options = new VertxOptions();
        // deploy(Vertx.vertx(options));
        options.setClusterManager(new HazelcastClusterManager());
        Vertx.clusteredVertx(options, res -> {
            deploy(res.result());
        });
    }

    public static void deploy(Vertx vertx) {
        vertx.deployVerticle(new ReceiveVerticle(), new DeploymentOptions().setInstances(1).setWorker(true), e -> {
            // when ReceiveVerticle is deployed, deploy SendVerticle
            vertx.deployVerticle(new SendVerticle(), new DeploymentOptions().setInstances(1).setWorker(true));
        });
    }

    public static class SendVerticle extends AbstractVerticle {
        @Override
        public void start() throws Exception {
            // start executing some job
            for (int i = 0; i < 5; i++) {
                // send some partial result asynchronously to ReceiveVerticle
                String message = String.valueOf(i);
                vertx.executeBlocking(f -> {
                    System.out.println("Sending message " + message);
                    vertx.eventBus().send("receive_address", message);
                    f.complete();
                }, false, r -> {
                });
                Thread.sleep(1000l);
            }
        }
    }

    public static class ReceiveVerticle extends AbstractVerticle {
        @Override
        public void start() throws Exception {
            vertx.eventBus().consumer("receive_address", e -> {
                System.out.println("Receive message " + e.body());
            });
        }
    }
```

Result is 
```
Sending message 0
Sending message 1
Sending message 2
Sending message 3
Sending message 4
Receive message 0
Receive message 1
Receive message 2
Receive message 3
Receive message 4
```

**Comment by tsegismont** [comment]
@gkozyryatskyy please see https://github.com/eclipse/vert.x/issues/1870#issuecomment-285727743

It is possible to send messages right away from executeblocking (we have a test for that), but not from a worker verticle.

**Comment by gkozyryatskyy** [comment]
@tsegismont so I need to create standard verticle and execute job in it in executeblocking?
Thank you for your reply!

**Comment by tsegismont** [comment]
You could do this yes. Keep in mind that executeBlocking and worker
verticles are not meant for executing infinite (or "almost" infinite) loops

If you need to loop like this, it's just fine to create a separate thread
(or a single thread executor) and publish to the event bus from there.

2017-05-19 15:38 GMT+02:00 gkozyryatskyy <notifications@github.com>:

**Comment by tsegismont** [comment]
@bauweto we had to fix an issue with respect to message ordering. When in clustered mode, sending implies handling an event telling the event bus which remote subscriber is going to handle the event. Since worker verticles can only handle one event at a time, the messages cannot be sent until the current event has been handled.

**Comment by bauweto** [comment]
@tsegismont thanks a lot for the answer. Can you point me to the issue you solved? Because if we use 3.4.0 then we might encounter the issue you fixed. 

From a functional point of view this behaviour looks strange. Image I get a million rows from a database with a worker verticle and for every row I send a message over the bus. The result is that the eventbus keeps on piling up rows until the worker verticle has finished retrieving all rows from the database. Aside from the memory consumption, this leads to a lot of negative implications. 

This difference between clustered and non clustered is just too big. It completely changes the premise of a vertx system. It takes you from a fast and asynchronous system back to a slow and sequential system.

Isn't there another solution for this?  For example, make the worker verticle handle more than 1 event but only for sending/processing messages?
I assume using a normal verticle with execute blocking is not a solution either because it will block the eventloop.

**Comment by tsegismont** [comment]
2017-08-05 13:12 GMT+02:00 Tom Bauwens <notifications@github.com>:

> @tsegismont <https://github.com/tsegismont> thanks a lot for the answer.
> Can you point me to the issue you solved? Because if we use 3.4.0 then we
> might encounter the issue you fixed.
>

There is no issue filed on GitHub, we found the bug looking into an
intermittent failure on CI. As said previously, we were no longer able to
guarantee the order of messages sent from the same context.


> From a functional point of view this behaviour looks strange. Image I get
> a million rows from a database with a worker verticle and for every row I
> send a message over the bus. The result is that the eventbus keeps on
> piling up rows until the worker verticle has finished retrieving all rows
> from the database. Aside from the memory consumption, this leads to a lot
> of negative implications.
>

Bear in mind that keeping a worker thread busy for such long tasks can lead
to all sorts of bad consequences. For example, networking or incoming event
bus message events could start to pile up as well.


> This difference between clustered and non clustered is just too big. It
> completely changes the premise of a vertx system. It takes you from a fast
> and asynchronous system back to a slow and sequential system.
>
> Isn't there another solution for this? For example, make the worker
> verticle handle more than 1 event but only for sending/processing messages?
> I assume using a normal verticle with execute blocking is not a solution
> either because it will block the eventloop.
>
No it would not block the event loop, executeBlocking is there precisely to
avoid that.

**Comment by tsegismont** [comment]
For your use case (very long running tasks), a separate WorkerExecutor
would be more appropriate.

2017-08-08 10:51 GMT+02:00 Thomas SEGISMONT <tsegismont@gmail.com>:
